### PR TITLE
Database refactor

### PR DESCRIPTION
**Pull request by Tymoteuszauke**
Added modules:
botpersistence,
models,
All test are ether ignored or commented (will be refactored)
database remodel

### CODE REVIEW

**Review by arturAR — APPROVED** [review]

File: botpersistence/src/main/java/com/blackbook/botpersistence/PersistenceApplication.java [hunk 1]
```diff
@@ -1,17 +1,13 @@
...
 import org.springframework.boot.web.servlet.ServletRegistrationBean;
 import org.springframework.context.annotation.Bean;
 
-/**
- *
- *
- */
 @SpringBootApplication
-public class CrawlerApp {
+public class PersistenceApplication {
```
Comment (mprtcz) [inline botpersistence/src/main/java/com/blackbook/botpersistence/PersistenceApplication.java]: Ugly naming :hankey:

File: botpersistence/src/main/java/com/blackbook/botpersistence/controller/PromotionsController.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    @Autowired
+    private PromotionsRepository promotionsRepository;
+
+    @Autowired
+    private PromotionParserService promotionParserService;
+
+    @Autowired
+    private AuthorsRepository authorsRepository;
+
+    @RequestMapping(method = RequestMethod.GET)
```
Comment (mprtcz) [inline botpersistence/src/main/java/com/blackbook/botpersistence/controller/PromotionsController.java]: Get is a default method :-1:

File: botpersistence/src/main/java/com/blackbook/botpersistence/controller/PromotionsController.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    public List<PromotionView> getPromotions(@RequestParam(defaultValue = "") String query,
+                                             @RequestParam(required = false) String priceFrom,
+                                             @RequestParam(required = false) String priceTo) {
+        log.info("Transaction: GET /api/promotions");
+
+        List<Promotion> promotions;
+
+        if (!StringUtils.isEmpty(priceFrom) && !StringUtils.isEmpty(priceTo)) {
```
Comment (mprtcz) [inline botpersistence/src/main/java/com/blackbook/botpersistence/controller/PromotionsController.java]: It's good to make a separate method for this

File: botpersistence/src/main/java/com/blackbook/botpersistence/dao/AuthorsRepository.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package com.blackbook.botpersistence.dao;
+
+
+import com.blackbook.botpersistence.model.Author;
```
Comment (mprtcz) [inline botpersistence/src/main/java/com/blackbook/botpersistence/dao/AuthorsRepository.java]: Better naming would be persistencebot

File: botpersistence/src/main/java/com/blackbook/botpersistence/dao/BooksRepository.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.repository.PagingAndSortingRepository;
+
+/**
+ * @author Siarhei Shauchenka
+ * @since 16.08.17
+ */
+public interface BooksRepository extends PagingAndSortingRepository<Book, Long> {
+
+//    @Query("SELECT b FROM Book b WHERE b.author.name LIKE %?1% OR b.title LIKE %?1%")
```
Comment (mprtcz) [inline botpersistence/src/main/java/com/blackbook/botpersistence/dao/BooksRepository.java]: :-1:

File: botpersistence/src/main/java/com/blackbook/botpersistence/service/PromotionParserService.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    @Autowired
+    private BookstoresRepository bookstoresRepository;
+
+    @Autowired
+    private AuthorsRepository authorsRepository;
+
+    @Autowired
+    private PromotionsRepository promotionsRepository;
+
+    /**
```
Comment (mprtcz) [inline botpersistence/src/main/java/com/blackbook/botpersistence/service/PromotionParserService.java]: :+1:

File: botpersistence/src/main/java/com/blackbook/botpersistence/service/PromotionParserService.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        promotion.setPrice(promotionCreationData.getPrice());
+        promotion.setPromotionDetails(promotionCreationData.getPromotionDetails());
+        Book parsedBook = parseBookCreationData(promotionCreationData.getBookCreationData());
+        Book book = booksRepository.findByTitle(parsedBook.getTitle());
+        if (book == null) {
+            book = parsedBook;
+        }
+        promotion.setBook(book);
+
+        //TODO remove if clause since creation data without bookstore id will not be permitted
```
Comment (mprtcz) [inline botpersistence/src/main/java/com/blackbook/botpersistence/service/PromotionParserService.java]: :-1:

File: botpersistence/src/main/java/com/blackbook/botpersistence/service/PromotionParserService.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Autowired
+    private PromotionsRepository promotionsRepository;
+
+    /**
+     * Parser API (from promotionCreationData into Promotion entity)
+     * Expected behavior is if promotion with given book id and given bookstore id is found, it deletes it and
+     * saves new promotion
+     */
+    public Promotion parsePromotionCreationData(PromotionCreationData promotionCreationData) {
```
Comment (mprtcz) [inline botpersistence/src/main/java/com/blackbook/botpersistence/service/PromotionParserService.java]: Maybe create a separate converter for those?

File: botpersistence/src/main/java/com/blackbook/botpersistence/service/PromotionParserService.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        //TODO remove if clause since creation data without bookstore id will not be permitted
+        if (promotionCreationData.getBookstoreId() != null) {
+            promotion.setBookstore(bookstoresRepository.findOne((long) promotionCreationData.getBookstoreId()));
+
+            Promotion existingPromotion = promotionsRepository.findByBookIdAndBookstoreId(book.getId(), promotionCreationData.getBookstoreId());
+            if (existingPromotion != null) {
+                promotionsRepository.delete(existingPromotion);
+            }
+        }
+
```
Comment (mprtcz) [inline botpersistence/src/main/java/com/blackbook/botpersistence/service/PromotionParserService.java]: Empty line :-1:

File: botpersistence/src/main/java/com/blackbook/botpersistence/util/ViewMapUtil.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+import view.promotion.BookAuthorView;
+import view.promotion.BookView;
+import view.promotion.BookstoreView;
+import view.promotion.PromotionView;
+
+import java.util.stream.Collectors;
+
+public class ViewMapUtil {
+
+    public static BookView BookViewFromBook(Book book) {
```
Comment (mprtcz) [inline botpersistence/src/main/java/com/blackbook/botpersistence/util/ViewMapUtil.java]: Why capitalized, still, ugly naming :-1:

File: botrest/src/main/java/com/blackbook/botrest/ApiCodes.java [hunk 11]
```diff
@@ -0,0 +1,7 @@
+package com.blackbook.botrest;
+
+public interface ApiCodes {
+
+    String PERSISTENCE_API = "http://localhost:8101";
```
Comment (mprtcz) [inline botrest/src/main/java/com/blackbook/botrest/ApiCodes.java]: Move it to properties file

File: botrest/src/test/resources/books_controller_test_dummy_data.sql [hunk 12]
```diff
@@ -1,4 +1,4 @@
-INSERT INTO books(author, title, price) VALUES('Tymke Wergiliusz', 'Emaks', 25.80);
-INSERT INTO books(author, title, price) VALUES('Tymek Robert', 'Skame', 80.80);
-INSERT INTO books(author, title, price) VALUES('Kamil Jurand', 'Scala dla topornych', 120.80);
-INSERT INTO books(author, title, price) VALUES('Macko Zdybko', 'Jak uszyc toporek', 49.99);
\ No newline at end of file
+--INSERT INTO books(author, title, price) VALUES('Tymke Wergiliusz', 'Emaks', 25.80);
+--INSERT INTO books(author, title, price) VALUES('Tymek Robert', 'Skame', 80.80);
+--INSERT INTO books(author, title, price) VALUES('Kamil Jurand', 'Scala dla topornych', 120.80);
+--INSERT INTO books(author, title, price) VALUES('Macko Zdybko', 'Jak usyzc toporek', 49.99);
```
Comment (mprtcz) [inline botrest/src/test/resources/books_controller_test_dummy_data.sql]: ? :suspect:

File: models/src/main/java/view/creation_model/BookCreationData.java [hunk 13]
```diff
@@ -0,0 +1,14 @@
+package view.creation_model;
+
+import lombok.Data;
+
+import java.util.List;
+
+
+@Data
+public class BookCreationData {
```
Comment (mprtcz) [inline models/src/main/java/view/creation_model/BookCreationData.java:9]: The name would rather be: BookDto -> Data Transfer Object

File: models/src/main/java/view/promotion/PromotionView.java [hunk 14]
```diff
@@ -0,0 +1,14 @@
+package view.promotion;
+
+import lombok.Data;
+
+@Data
+public class PromotionView {
+
```
Comment (mprtcz) [inline models/src/main/java/view/promotion/PromotionView.java:7]: :-1: empty line

**Review by mprtcz — APPROVED** [review]

**Review by patkkka — APPROVED** [review]

**Pull request merged**